boards/shared/Src/Io/Io_SharedFreqOnlyPwmInput.c

```diff
@@ -0,0 +1,155 @@
+    uint32_t              timer_auto_reload_reg;
+    HAL_TIM_ActiveChannel timer_active_channel;
+
+    bool     first_tick;
+    uint32_t curr_rising_edge;
+    uint32_t prev_rising_edge;
+
+    size_t timer_overflow_count;
+};
+
+/**
+ * Set the frequency for the given PWM input
+ * @param pwm_input: The PWM input to set frequency for
+ * @param frequency_hz: Frequency, in Hz
+ */
+static void Io_SetFrequency(
+    struct FreqOnlyPwmInput *pwm_input,
+    const float              frequency_hz);
+
+static void Io_SetFrequency(
+    struct FreqOnlyPwmInput *const pwm_input,
+    const float                    frequency_hz)
+{
+    assert(frequency_hz >= 0.0f || isnanf(frequency_hz));
+
+    pwm_input->frequency_hz = frequency_hz;
+}
+
+struct FreqOnlyPwmInput *Io_SharedFreqOnlyPwmInput_Create(
+    TIM_HandleTypeDef *   htim,
+    float                 timer_frequency_hz,
+    uint32_t              rising_edge_timer_channel,
+    uint32_t              timer_auto_reload_reg,
+    HAL_TIM_ActiveChannel timer_active_channel)
+{
+    assert(htim != NULL);
+    static struct FreqOnlyPwmInput pwm_inputs[MAX_NUM_OF_FREQ_ONLY_PWM_INPUTS];
+    static size_t                  alloc_index = 0;
+
+    assert(alloc_index < MAX_NUM_OF_FREQ_ONLY_PWM_INPUTS);
+
+    struct FreqOnlyPwmInput *const pwm_input = &pwm_inputs[alloc_index++];
+
+    pwm_input->frequency_hz            = NAN;
+    pwm_input->htim                    = htim;
+    pwm_input->timer_frequency_hz      = timer_frequency_hz;
+    pwm_input->rising_edge_tim_channel = rising_edge_timer_channel;
+    pwm_input->timer_auto_reload_reg   = timer_auto_reload_reg;
+    pwm_input->timer_active_channel    = timer_active_channel;
+    pwm_input->prev_rising_edge        = 0U;
+    pwm_input->curr_rising_edge        = 0U;
+    pwm_input->first_tick              = true;
+    pwm_input->timer_overflow_count    = 0U;
+
+    HAL_TIM_IC_Start_IT(htim, rising_edge_timer_channel);
+    HAL_TIM_Base_Start_IT(htim);
+
+    return pwm_input;
+}
+
+float Io_SharedFreqOnlyPwmInput_GetFrequency(
+    struct FreqOnlyPwmInput *const pwm_input)
+{
+    return pwm_input->frequency_hz;
+}
+
+TIM_HandleTypeDef *Io_SharedFreqOnlyPwmInput_GetTimerHandle(
+    struct FreqOnlyPwmInput *const pwm_input)
+{
+    return pwm_input->htim;
+}
+
+HAL_TIM_ActiveChannel Io_SharedFreqOnlyPwmInput_GetTimerActiveChannel(
+    struct FreqOnlyPwmInput *const pwm_input)
+{
+    return pwm_input->timer_active_channel;
+}
+
+void Io_SharedFreqOnlyPwmInput_Tick(struct FreqOnlyPwmInput *const pwm_input)
+{
+    // Reset elapsed tick counter for the flow meter
+    pwm_input->timer_overflow_count = 0U;
+
+    /**
+     * @note The function captures the value of first rising edge, and then
+     * captures the second rising edge from the capture compare unit. The
+     * difference between these two values, rising_edge_delta are computed and
+     * used to compute the frequency of the pwm input signal
+     */
```

Comment:
```suggestion
    // We store the counter values captured during two most recent rising edges.
    // The difference between these two counter values is used to compute the
    // frequency of the PWM input signal.
```
nit; wording and block comment style